src/sync/scheduler.test.ts

```diff
@@ -0,0 +1,221 @@
+		subscription = ret.subscription;
+		const repoSyncState = ret.repoSyncState!;
+		for (let newRepoStateNo = 1; newRepoStateNo < 500; newRepoStateNo++) {
+			const plainObj = repoSyncState.get();
+			delete plainObj["id"];
+			delete plainObj["commitStatus"];
+			delete plainObj["branchStatus"];
+			const newRepoSyncState = new RepoSyncState(plainObj);
+			newRepoSyncState.repoId = repoSyncState.repoId + newRepoStateNo;
+			newRepoSyncState.repoName = repoSyncState.repoName + newRepoStateNo;
+			newRepoSyncState.repoFullName = repoSyncState.repoFullName + newRepoStateNo;
+			await newRepoSyncState.save();
+		}
+		when(booleanFlag).calledWith(
+			BooleanFlags.USE_SUBTASKS_FOR_BACKFILL,
+			expect.anything()
+		).mockResolvedValue(false);
+	});
+
+	const configureRateLimit = (coreQuotaRemainig: number, graphQlQuotaRemaining: number) => {
+		githubNock
+			.persist()
+			.get(`/rate_limit`)
+			.reply(200, {
+				"resources": {
+					"core": {
+						"limit": coreQuotaRemainig * 10,
+						"remaining": coreQuotaRemainig
+					},
+					"graphql": {
+						"limit": graphQlQuotaRemaining * 10,
+						"remaining": graphQlQuotaRemaining
+					}
+				}
+			});
+	};
+
+	it("always returns repo task if not complete yet", async () => {
+		subscription.repositoryStatus = undefined;
+		const nextTasks = await getNextTasks(subscription, [], getLogger("test"));
+		expect(nextTasks[0].task).toEqual("repository");
+	});
+
+	it("first (main) task is always same (deterministic), when FF off", async () => {
+		const firstExecutionResult = await getNextTasks(subscription, [], getLogger("test"));
+
+		for (let i = 0; i < 10; i++) {
+			const nextExecutionResult = await getNextTasks(subscription, [], getLogger("test"));
+			expect(nextExecutionResult).toEqual(firstExecutionResult);
+		}
+	});
+
+	it("first (main) task is always same (deterministic), when FF on", async () => {
+		when(booleanFlag).calledWith(
+			BooleanFlags.USE_SUBTASKS_FOR_BACKFILL,
+			expect.anything()
+		).mockResolvedValue(true);
+
+		configureRateLimit(10000, 10000);
+
+		githubUserTokenNock(DatabaseStateCreator.GITHUB_INSTALLATION_ID);
+		const firstTask = (await getNextTasks(subscription, [], getLogger("test")))[0];
+
+		for (let i = 0; i < 10; i++) {
+			const tasks = await getNextTasks(subscription, [], getLogger("test"));
+			expect(tasks[0]).toEqual(firstTask);
+		}
+	});
+
+	it("returns only main task when FF is off", async () => {
+		expect((await getNextTasks(subscription, [], getLogger("test"))).length).toEqual(1);
+	});
+
+	it("uses smallest quota between core and graphql to determine number of subtasks", async () => {
+		when(booleanFlag).calledWith(
+			BooleanFlags.USE_SUBTASKS_FOR_BACKFILL,
+			expect.anything()
+		).mockResolvedValue(true);
+
+		configureRateLimit(100000, 2000);
+
+		githubUserTokenNock(DatabaseStateCreator.GITHUB_INSTALLATION_ID);
+		const tasks = await getNextTasks(subscription, [], getLogger("test"));
+		expect(tasks.length).toEqual(4);
```

Comment:
/nit
I dont' get where does the `4` comes from within this test, unless I dig into the impl code, which I don't want to.